server.js

```diff
@@ -1,29 +1,119 @@
-mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true })
+// eslint-disable-next-line max-len
+mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true, useCreateIndex: true })
 mongoose.Promise = Promise
 
-// Defines the port the app will run on. Defaults to 8080, but can be 
-// overridden when starting the server. For example:
-//
-//   PORT=9000 npm start
-const port = process.env.PORT || 8080
+const thougthSchema = new mongoose.Schema({
+  message: {
+    type: String,
+    required: [true, 'Empty message not allowed'],
+    trim: true,
+    minlength: [5, 'Message must be longer than 5 characters'],
+    maxlength: [140, 'Message must be shorter than 140 characters']
+  },
+  hearts: {
+    type: Number,
+    default: 0
+  },
+  createdAt: {
+    type: Date,
+    default: Date.now
+  }
+})
+
+const Thought = mongoose.model('Thought', thougthSchema)
+
+const port = process.env.PORT || 8081
 const app = express()
 
-// Add middlewares to enable cors and json body parsing
 app.use(cors())
-app.use(bodyParser.json())
+app.use(express.json())
 
-// Start defining your routes here
 app.get('/', (req, res) => {
-  res.send('Hello world')
+  res.send(listEndpoints(app))
+})
+
+app.get('/thoughts', async (req, res) => {
+  const allThoughts = await Thought.find().sort({ createdAt: -1 }).limit(20).exec()
+  
+  res.json(allThoughts)
+})
+
+app.get('/thoughts/:id', async (req, res) => {
+  const { id } = req.params
+
+  const thought = await Thought.findById(id)
+  res.json(thought)
+})
+
+// POST endpoint for new thought
+app.post('/thoughts', async (req, res) => {
+  try {
+    const newThought = await new Thought({ message: req.body.message }).save()
+    res.json(newThought)
+  } catch (error) {
+    if (error.code === 11000) {
+      res.status(400).json({ error: 'Duplicated value', fields: error.keyValue })
+    }
+    res.status(400).json(error)
+  }
+})
+
+// POST endpoint to update number of hearts on thought
+app.post('/thoughts/:id/likes', async (req, res) => {
+  const { id } = req.params
+
+  try {
+    const updatedThought = await Thought.findOneAndUpdate(
+      { 
+        _id: id 
+      }, 
+      { 
+        $inc: 
+          { 
+            hearts: 1 
+          } 
+      }, 
+      { 
+        new: true 
+      }
+    )
+
+    if (updatedThought) {
+      res.json(updatedThought)
+    } else {
+      res.status(404).json({ message: 'Not found' })
+    }
+  } catch (error) {
+    res.status(400).json({ message: 'Invalid request', error })
+  }
+})
+
+// DELETE endpoint to find and delete a thought
+app.delete('/thoughts/:id', async (req, res) => {
+  const { id } = req.params
+
+  try {
+    const deletedThought = await Thought.findOneAndDelete({ _id: id })
+    if (deletedThought) {
+      res.json(deletedThought)
+    } else {
+      res.status(404).json({ message: 'Not found' })
+    }
+  } catch (error) {
+    res.status(400).json({ message: 'Invalid request', error })
+  }
 })
```

Comment:
Remember to restrict access to this endpoint because not all users should have access to this functionality 🤓